### DISCUSSION TITLE
IDbContext and repository pattern

### DISCUSSION THREAD

**Discussion by BartoGabriel** [jasontaylordev/CleanArchitecture#620] [Q&A]
First, thanks for this repository. It is of great help.

I have some questions regarding the IDbContext and repository pattern.
The Project uses IDbContext and it feels very comfortable to work directly with the EfCore abstraction. It is very useful for the vast majority of use cases.
But in some cases, I have quite advanced and specific SQL queries. I think those deserve to have an application layer abstraction.

Which leaves me with a few questions:
1. Would it be correct to use the repository pattern only for certain complex queries? I imagine having the repository Interface in the application layer, and the implementation in Infrastructure.
2. Since we have some complex queries, in some specific cases we use Dapper, since in some cases it is much more versatile than EF. With which separating those cases in a specific IRepository would allow me in the infrastructure layer to select the framework for the specific use case.
This doubt arises, because in most cases the IDbContext is enough for me.

On the other hand, something I like about the repository pattern is having the aggregate entities identified, and that they only access the parent and not the children. Which brings up the third question:

3. In the pattern used by this project, can the aggregate entities be identified so that the children are not handled in isolation?

Thanks!

**Comment by neman** [comment]
Repository interface is part of domain model in ddd and clean architecture

**Reply by BartoGabriel** [reply]
Thank you very much for the reply. I would like to know what you think about mixing the architecture of this repository with DDD.

I feel very comfortable for most cases, but many times I like that certain logic is in the domain, and that the parent entity is responsible for the children (Orders -> OrderItem), by not having aggregate this it is free to that anyone can modify OrderItem and control is lost.

**Reply by BartoGabriel** [reply]
I usually put repository interfaces in the application layer. It often feels more comfortable.
I know I'm not the only one who does.

Example:
https://github.com/jasontaylordev/NorthwindTraders/issues/130#issuecomment-502075566
